From 1e379e04c3fc1457a80d47e0da2a6e242237094c Mon Sep 17 00:00:00 2001
From: Karol Herbst <kherbst@redhat.com>
Date: Wed, 1 Apr 2026 02:55:52 +0200
Subject: [PATCH] libclc: Partly revert "libclc: Update trigpi functions
 (#187579)"

The initial commit broke sinpi

This partly reverts commit 421bf13e4bf1c19a0cc48fbb0af63b902d3118b1

Upstream: https://gitlab.freedesktop.org/karolherbst/mesa-libclc/-/commit/1e379e04c3fc1457a80d47e0da2a6e242237094c

Signed-off-by: Bernd Kuhls <bernd@kuhls.net>
---
 .../clc/math/clc_sincos_helpers_fp64_decl.inc |   5 +
 libclc/clc/lib/generic/math/clc_sinpi.cl             |   8 +-
 libclc/clc/lib/generic/math/clc_sinpi.inc            | 106 +++++++++++++++++-
 3 files changed, 115 insertions(+), 4 deletions(-)

diff --git a/libclc/clc/include/clc/math/clc_sincos_helpers_fp64_decl.inc b/libclc/clc/include/clc/math/clc_sincos_helpers_fp64_decl.inc
index 22081d9..d638d72 100644
--- a/libclc/clc/include/clc/math/clc_sincos_helpers_fp64_decl.inc
+++ b/libclc/clc/include/clc/math/clc_sincos_helpers_fp64_decl.inc
@@ -22,6 +22,11 @@ _CLC_DEF _CLC_OVERLOAD __CLC_DOUBLEN __clc_tan_reduced_eval(__CLC_DOUBLEN x,
                                                             __CLC_DOUBLEN y,
                                                             __CLC_INTN is_odd);
 
+_CLC_DECL _CLC_OVERLOAD void __clc_sincos_piby4(__CLC_DOUBLEN x,
+                                                __CLC_DOUBLEN xx,
+                                                private __CLC_DOUBLEN *sinval,
+                                                private __CLC_DOUBLEN *cosval);
+
 _CLC_DECL _CLC_OVERLOAD __CLC_INTN __clc_remainder_piby2_small(
     __CLC_DOUBLEN x, private __CLC_DOUBLEN *r, private __CLC_DOUBLEN *rr);
 
diff --git a/libclc/clc/lib/generic/math/clc_sinpi.cl b/libclc/clc/lib/generic/math/clc_sinpi.cl
index bf57dc4..f9bed7e 100644
--- a/libclc/clc/lib/generic/math/clc_sinpi.cl
+++ b/libclc/clc/lib/generic/math/clc_sinpi.cl
@@ -6,8 +6,12 @@
 //
 //===----------------------------------------------------------------------===//
 
-#include "clc/math/clc_sincospi.h"
-#include "clc/math/clc_sinpi.h"
+#include "clc/clc_convert.h"
+#include "clc/float/definitions.h"
+#include "clc/internal/clc.h"
+#include "clc/math/clc_fabs.h"
+#include "clc/math/clc_sincos_helpers.h"
+#include "clc/math/math.h"
 
 #define __CLC_BODY "clc_sinpi.inc"
 #include "clc/math/gentype.inc"
diff --git a/libclc/clc/lib/generic/math/clc_sinpi.inc b/libclc/clc/lib/generic/math/clc_sinpi.inc
index a6e5ee7..264609a 100644
--- a/libclc/clc/lib/generic/math/clc_sinpi.inc
+++ b/libclc/clc/lib/generic/math/clc_sinpi.inc
@@ -6,7 +6,109 @@
 //
 //===----------------------------------------------------------------------===//
 
+#if __CLC_FPSIZE == 32
+
 _CLC_OVERLOAD _CLC_DEF __CLC_GENTYPE __clc_sinpi(__CLC_GENTYPE x) {
-  __CLC_GENTYPE unused_cos;
-  return __clc_sincospi(x, &unused_cos);
+  __CLC_INTN ix = __CLC_AS_INTN(x);
+  __CLC_INTN xsgn = ix & (__CLC_INTN)0x80000000;
+  ix ^= xsgn;
+  __CLC_GENTYPE absx = __clc_fabs(x);
+  __CLC_INTN iax = __CLC_CONVERT_INTN(absx);
+  __CLC_GENTYPE r = absx - __CLC_CONVERT_GENTYPE(iax);
+  __CLC_INTN xodd =
+      xsgn ^ ((iax & 0x1) != 0 ? (__CLC_INTN)0x80000000 : (__CLC_INTN)0);
+
+  // Initialize with return for +-Inf and NaN
+  __CLC_INTN ir = QNANBITPATT_SP32;
+
+  // 2^23 <= |x| < Inf, the result is always integer
+  ir = ix < PINFBITPATT_SP32 ? xsgn : ir;
+
+  // 0x1.0p-7 <= |x| < 2^23, result depends on which 0.25 interval
+
+  // r < 1.0
+  __CLC_GENTYPE a = 1.0f - r;
+  __CLC_INTN e = 0;
+
+  // r <= 0.75
+  __CLC_INTN c = r <= 0.75f;
+  a = c ? r - 0.5f : a;
+  e = c ? 1 : e;
+
+  // r < 0.5
+  c = r < 0.5f;
+  a = c ? 0.5f - r : a;
+
+  // 0 < r <= 0.25
+  c = r <= 0.25f;
+  a = c ? r : a;
+  e = c ? 0 : e;
+
+  __CLC_GENTYPE sinval, cosval;
+  __clc_sincos_piby4(a * M_PI_F, &sinval, &cosval);
+  __CLC_INTN jr = xodd ^ __CLC_AS_INTN(e != 0 ? cosval : sinval);
+
+  ir = ix < 0x4b000000 ? jr : ir;
+
+  return __CLC_AS_GENTYPE(ir);
 }
+
+#elif __CLC_FPSIZE == 64
+
+_CLC_OVERLOAD _CLC_DEF __CLC_GENTYPE __clc_sinpi(__CLC_GENTYPE x) {
+  __CLC_LONGN ix = __CLC_AS_LONGN(x);
+  __CLC_LONGN xsgn = ix & (__CLC_LONGN)0x8000000000000000L;
+  ix ^= xsgn;
+  __CLC_GENTYPE absx = __clc_fabs(x);
+  __CLC_LONGN iax = __CLC_CONVERT_LONGN(absx);
+  __CLC_GENTYPE r = absx - __CLC_CONVERT_GENTYPE(iax);
+  __CLC_LONGN xodd =
+      xsgn ^
+      ((iax & 0x1L) != 0 ? (__CLC_LONGN)0x8000000000000000L : (__CLC_LONGN)0L);
+
+  // Initialize with return for +-Inf and NaN
+  __CLC_LONGN ir = QNANBITPATT_DP64;
+
+  // 2^23 <= |x| < Inf, the result is always integer
+  ir = ix < PINFBITPATT_DP64 ? xsgn : ir;
+
+  // 0x1.0p-7 <= |x| < 2^23, result depends on which 0.25 interval
+
+  // r < 1.0
+  __CLC_GENTYPE a = 1.0 - r;
+  __CLC_LONGN e = 0;
+
+  //  r <= 0.75
+  __CLC_LONGN c = r <= 0.75;
+  __CLC_GENTYPE t = r - 0.5;
+  a = c ? t : a;
+  e = c ? 1 : e;
+
+  // r < 0.5
+  c = r < 0.5;
+  t = 0.5 - r;
+  a = c ? t : a;
+
+  // r <= 0.25
+  c = r <= 0.25;
+  a = c ? r : a;
+  e = c ? 0 : e;
+
+  __CLC_GENTYPE api = a * M_PI;
+
+  __CLC_GENTYPE sinval, cosval;
+  __clc_sincos_piby4(api, 0.0, &sinval, &cosval);
+  __CLC_LONGN jr = xodd ^ __CLC_AS_LONGN(e != 0 ? cosval : sinval);
+
+  ir = absx < 0x1.0p+52 ? jr : ir;
+
+  return __CLC_AS_GENTYPE(ir);
+}
+
+#elif __CLC_FPSIZE == 16
+
+_CLC_OVERLOAD _CLC_DEF __CLC_GENTYPE __clc_sinpi(__CLC_GENTYPE x) {
+  return __CLC_CONVERT_GENTYPE(__clc_sinpi(__CLC_CONVERT_FLOATN(x)));
+}
+
+#endif
-- 
2.47.3

